File: internal/github/repos.go
package github

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"sort"
	"strings"
)

// Repo represents a GitHub repository from the API.
type Repo struct {
	ID            int64    `json:"id"`
	NodeID        string   `json:"node_id"`
	Name          string   `json:"name"`
	FullName      string   `json:"full_name"`
	Description   *string  `json:"description"`
	Private       bool     `json:"private"`
	Owner         Owner    `json:"owner"`
	HTMLURL       string   `json:"html_url"`
	CloneURL      string   `json:"clone_url"`
	Language      *string  `json:"language"`
	StargazersCount int    `json:"stargazers_count"`
	ForksCount    int      `json:"forks_count"`
	DefaultBranch string   `json:"default_branch"`
	Topics        []string `json:"topics"`
	PushedAt      string   `json:"pushed_at"`
	CreatedAt     string   `json:"created_at"`
	UpdatedAt     string   `json:"updated_at"`
	Archived      bool     `json:"archived"`
	Fork          bool     `json:"fork"`
	Size          int      `json:"size"`
}

// Owner represents a GitHub repository owner.
type Owner struct {
	Login string `json:"login"`
	ID    int64  `json:"id"`
}

// readmeResponse represents the API response for README requests.
type readmeResponse struct {
	Content  string `json:"content"`
	Encoding string `json:"encoding"`
}

// treeResponse represents the API response for tree requests.
type treeResponse struct {
	Tree []treeEntry `json:"tree"`
}

type treeEntry struct {
	Path string `json:"path"`
	Mode string `json:"mode"`
	Type string `json:"type"` // "blob" or "tree"
	SHA  string `json:"sha"`
	Size int    `json:"size,omitempty"`
}

// FetchOwnedRepos returns a channel that yields owned repos one at a time.
// Pagination is handled internally. The channel is closed when all pages
// are exhausted or an error occurs. Errors are sent on the errCh channel.
func FetchOwnedRepos(ctx context.Context, client *Client) (<-chan Repo, <-chan error) {
	repoCh := make(chan Repo, 100)
	errCh := make(chan error, 1)

	go func() {
		defer close(repoCh)
		defer close(errCh)

		page := 1
		for {
			path := fmt.Sprintf("/user/repos?type=owner&sort=pushed&direction=desc&per_page=100&page=%d", page)
			resp, err := client.Get(ctx, path)
			if err != nil {
				errCh <- err
				return
			}

			body, err := io.ReadAll(resp.Body)
			_ = resp.Body.Close()
			if err != nil {
				errCh <- err
				return
			}

			if resp.StatusCode != 200 {
				errCh <- fmt.Errorf("GitHub API error: %s", resp.Status)
				return
			}

			var repos []Repo
			if err := json.Unmarshal(body, &repos); err != nil {
				errCh <- err
				return
			}

			if len(repos) == 0 {
				return
			}

			for _, repo := range repos {
				select {
				case repoCh <- repo:
				case <-ctx.Done():
					errCh <- ctx.Err()
					return
				}
			}

			if len(repos) < 100 {
				return
			}

			page++
		}
	}()

	return repoCh, errCh
}

// FetchStarredRepos returns a channel that yields starred repos one at a time.
func FetchStarredRepos(ctx context.Context, client *Client) (<-chan Repo, <-chan error) {
	repoCh := make(chan Repo, 100)
	errCh := make(chan error, 1)

	go func() {
		defer close(repoCh)
		defer close(errCh)

		page := 1
		for {
			path := fmt.Sprintf("/user/starred?per_page=100&page=%d", page)
			resp, err := client.Get(ctx, path)
			if err != nil {
				errCh <- err
				return
			}

			body, err := io.ReadAll(resp.Body)
			_ = resp.Body.Close()
			if err != nil {
				errCh <- err
				return
			}

			if resp.StatusCode != 200 {
				errCh <- fmt.Errorf("GitHub API error: %s", resp.Status)
				return
			}

			var repos []Repo
			if err := json.Unmarshal(body, &repos); err != nil {
				errCh <- err
				return
			}

			if len(repos) == 0 {
				return
			}

			for _, repo := range repos {
				select {
				case repoCh <- repo:
				case <-ctx.Done():
					errCh <- ctx.Err()
					return
				}
			}

			if len(repos) < 100 {
				return
			}

			page++
		}
	}()

	return repoCh, errCh
}

// FetchReadme fetches and base64-decodes the README for the given repo.
// Returns ("", nil) if the repo has no README (404).
// Never returns an error — logs and returns empty string on any failure.
func FetchReadme(ctx context.Context, client *Client, owner, repo string) string {
	path := fmt.Sprintf("/repos/%s/%s/readme", owner, repo)
	resp, err := client.Get(ctx, path)
	if err != nil {
		return ""
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode == 404 {
		return ""
	}

	if resp.StatusCode != 200 {
		return ""
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return ""
	}

	var readme readmeResponse
	if err := json.Unmarshal(body, &readme); err != nil {
		return ""
	}

	if readme.Encoding != "base64" || readme.Content == "" {
		return ""
	}

	// Strip newlines before decoding
	content := strings.ReplaceAll(readme.Content, "\n", "")
	decoded, err := base64.StdEncoding.DecodeString(content)
	if err != nil {
		return ""
	}

	return strings.TrimRight(string(decoded), "\n\r")
}

// FetchFileTree fetches the git tree for the given repo and branch.
// Returns only blob entries at depth <= 2 (paths with at most one '/'),
// sorted alphabetically, joined by newlines.
// Returns "" on any error — never returns an error.
func FetchFileTree(ctx context.Context, client *Client, owner, repo, branch string) string {
	path := fmt.Sprintf("/repos/%s/%s/git/trees/%s?recursive=1", owner, repo, branch)
	resp, err := client.Get(ctx, path)
	if err != nil {
		return ""
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != 200 {
		return ""
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return ""
	}

	var tree treeResponse
	if err := json.Unmarshal(body, &tree); err != nil {
		return ""
	}

	var paths []string
	for _, entry := range tree.Tree {
		// Only include blobs (files)
		if entry.Type != "blob" {
			continue
		}

		// Filter depth <= 2 (at most one '/' in the path)
		slashCount := strings.Count(entry.Path, "/")
		if slashCount > 1 {
			continue
		}

		paths = append(paths, entry.Path)
	}

	if len(paths) == 0 {
		return ""
	}

	// Sort alphabetically
	sort.Strings(paths)

	return strings.Join(paths, "\n")
}

File: _legacy/packages/core/src/github/repos.ts
/**
 * GitHub API data fetching module.
 *
 * Provides functions to fetch repository data from GitHub's REST API,
 * including owned repos, starred repos, READMEs, and file trees.
 */

import { GitHubClient, withRateLimitRetry } from './client.js';

export interface GitHubRepo {
  id: number;
  node_id: string;
  name: string;
  full_name: string;
  description: string | null;
  private: boolean;
  owner: {
    login: string;
    id: number;
  };
  html_url: string;
  clone_url: string;
  ssh_url: string;
  language: string | null;
  stargazers_count: number;
  forks_count: number;
  open_issues_count: number;
  default_branch: string;
  topics: string[];
  pushed_at: string;
  created_at: string;
  updated_at: string;
  archived: boolean;
  fork: boolean;
  size: number;
}

export interface FetchReposOptions {
  owned?: boolean;
  starred?: boolean;
}

/**
 * Fetch owned repositories, paginating until done.
 *
 * @param client - Authenticated GitHub client
 * @returns Async generator yielding GitHubRepo objects
 */
export async function* fetchOwnedRepos(
  client: GitHubClient
): AsyncGenerator<GitHubRepo> {
  let page = 1;
  const perPage = 100;

  while (true) {
    try {
      const response = await withRateLimitRetry(() =>
        client.rest.repos.listForAuthenticatedUser({
          visibility: 'all',
          sort: 'pushed',
          direction: 'desc',
          per_page: perPage,
          page,
        })
      );

      const repos = response.data;

      if (!repos || repos.length === 0) {
        break;
      }

      for (const repo of repos) {
        // Cast strictly to our interface to ensure we have the fields we expect
        yield repo as unknown as GitHubRepo;
      }

      if (repos.length < perPage) {
        break;
      }

      page++;
    } catch (error) {
      console.error(`Error fetching owned repos page ${page}:`, error);
      // Stop pagination on error, but don't crash the whole process
      break;
    }
  }
}

/**
 * Fetch starred repositories, paginating until done.
 *
 * @param client - Authenticated GitHub client
 * @returns Async generator yielding GitHubRepo objects
 */
export async function* fetchStarredRepos(
  client: GitHubClient
): AsyncGenerator<GitHubRepo> {
  let page = 1;
  const perPage = 100;

  while (true) {
    try {
      const response = await withRateLimitRetry(() =>
        client.rest.activity.listReposStarredByAuthenticatedUser({
          sort: 'created', // standard sort for starred
          direction: 'desc',
          per_page: perPage,
          page,
        })
      );

      const repos = response.data;

      if (!repos || repos.length === 0) {
        break;
      }

      for (const repo of repos) {
        yield repo as unknown as GitHubRepo;
      }

      if (repos.length < perPage) {
        break;
      }

      page++;
    } catch (error) {
      console.error(`Error fetching starred repos page ${page}:`, error);
      break;
    }
  }
}

/**
 * Fetch the README content for a repository.
 *
 * @param client - Authenticated GitHub client
 * @param owner - Repository owner
 * @param repo - Repository name
 * @returns Decoded README content or null if not found
 */
export async function fetchReadme(
  client: GitHubClient,
  owner: string,
  repo: string
): Promise<string | null> {
  try {
    const response = await withRateLimitRetry(() =>
      client.rest.repos.getReadme({
        owner,
        repo,
      })
    );

    const content = response.data.content;
    if (!content) {
      return null;
    }

    const decoded = Buffer.from(content, 'base64').toString('utf-8');
    return decoded.trimEnd();
  } catch (error: unknown) {
    // 404 is expected for repos without README
    const httpError = error as { status?: number };
    if (httpError.status === 404) {
      return null;
    }
    // Log other errors but don't throw
    console.error(`Error fetching README for ${owner}/${repo}:`, error instanceof Error ? error.message : error);
    return null;
  }
}

/**
 * Fetch the file tree for a repository.
 *
 * @param client - Authenticated GitHub client
 * @param owner - Repository owner
 * @param repo - Repository name
 * @param branch - Branch to fetch tree for (usually default branch)
 * @returns List of file paths separated by newlines, or null if error
 */
export async function fetchFileTree(
  client: GitHubClient,
  owner: string,
  repo: string,
  branch: string
): Promise<string | null> {
  try {
    const response = await withRateLimitRetry(() =>
      client.rest.git.getTree({
        owner,
        repo,
        tree_sha: branch,
        recursive: '1',
      })
    );

    const tree = response.data.tree;
    if (!tree) {
      return null;
    }

    const filePaths = tree
      .filter((item) => {
        // Only include blobs (files)
        if (item.type !== 'blob' || !item.path) {
          return false;
        }
        
        // Filter depth <= 2 (at most one slash)
        // e.g. "README.md" (0 slashes) -> OK
        // e.g. "src/index.ts" (1 slash) -> OK
        // e.g. "src/components/Button.tsx" (2 slashes) -> Skip
        const slashCount = (item.path.match(/\//g) || []).length;
        return slashCount <= 1;
      })
      .map((item) => item.path!)
      .sort();

    if (filePaths.length === 0) {
      return null;
    }

    return filePaths.join('\n');
  } catch (error: unknown) {
    console.error(`Error fetching file tree for ${owner}/${repo}:`, error instanceof Error ? error.message : error);
    return null;
  }
}
